File: _plugins/academic_numbering.rb
# frozen_string_literal: true

# Academic Numbering System for Jekyll
# 
# This plugin provides LaTeX-like automatic numbering and cross-referencing
# for figures, tables, equations, and algorithms in blog posts.
# 
# Usage:
#   Figures:
#     {% figure id:fig-example caption:"An example figure" %}
#       <img src="..." alt="...">
#     {% endfigure %}
#     
#     Reference: {% ref fig-example %}  -> Figure 1
#   
#   Tables:
#     {% table id:tab-example caption:"An example table" %}
#       | Header 1 | Header 2 |
#       |----------|----------|
#       | Cell 1   | Cell 2   |
#     {% endtable %}
#     
#     Reference: {% ref tab-example %}  -> Table 1
#   
#   Equations (labeled):
#     {% equation id:eq-example %}
#       E = mc^2
#     {% endequation %}
#     
#     Reference: {% ref eq-example %}  -> Eq. 1
#   
#   Algorithms:
#     {% algorithm id:alg-example caption:"My Algorithm" %}
#       ...algorithm content...
#     {% endalgorithm %}
#     
#     Reference: {% ref alg-example %}  -> Algorithm 1

module Jekyll
  # Store for numbering counters (per-page)
  module AcademicNumbering
    def self.reset_counters
      @counters = { 'figure' => 0, 'table' => 0, 'equation' => 0, 'algorithm' => 0 }
      @labels = {}
    end
    
    def self.increment(type)
      @counters ||= { 'figure' => 0, 'table' => 0, 'equation' => 0, 'algorithm' => 0 }
      @counters[type] = (@counters[type] || 0) + 1
    end
    
    def self.current(type)
      @counters ||= { 'figure' => 0, 'table' => 0, 'equation' => 0, 'algorithm' => 0 }
      @counters[type] || 0
    end
    
    def self.register_label(id, type, number)
      @labels ||= {}
      @labels[id] = { type: type, number: number }
    end
    
    def self.get_label(id)
      @labels ||= {}
      @labels[id]
    end
    
    def self.labels
      @labels ||= {}
    end
  end

  # Figure block tag: {% figure id:fig-id caption:"Caption text" %}...{% endfigure %}
  class FigureBlock < Liquid::Block
    def initialize(tag_name, markup, tokens)
      super
      @id = nil
      @caption = nil
      
      # Parse id:value and caption:"value"
      if markup =~ /id:\s*([^\s]+)/
        @id = $1.strip
      end
      if markup =~ /caption:\s*"([^"]+)"/
        @caption = $1.strip
      elsif markup =~ /caption:\s*'([^']+)'/
        @caption = $1.strip
      end
    end

    def render(context)
      content = super.to_s.strip
      number = AcademicNumbering.increment('figure')
      
      if @id
        AcademicNumbering.register_label(@id, 'figure', number)
      end
      
      id_attr = @id ? " id=\"#{@id}\"" : ""
      caption_html = @caption ? "<figcaption class=\"figure-caption\"><strong>Figure #{number}:</strong> #{@caption}</figcaption>" : "<figcaption class=\"figure-caption\"><strong>Figure #{number}</strong></figcaption>"
      
      <<~HTML
        <figure class="academic-figure"#{id_attr} data-figure-number="#{number}">
          <div class="figure-content">
            #{content}
          </div>
          #{caption_html}
        </figure>
      HTML
    end
  end

  # Table block tag: {% table id:tab-id caption:"Caption text" %}...{% endtable %}
  class TableBlock < Liquid::Block
    def initialize(tag_name, markup, tokens)
      super
      @id = nil
      @caption = nil
      
      if markup =~ /id:\s*([^\s]+)/
        @id = $1.strip
      end
      if markup =~ /caption:\s*"([^"]+)"/
        @caption = $1.strip
      elsif markup =~ /caption:\s*'([^']+)'/
        @caption = $1.strip
      end
    end

    def render(context)
      content = super.to_s.strip
      number = AcademicNumbering.increment('table')
      
      if @id
        AcademicNumbering.register_label(@id, 'table', number)
      end
      
      id_attr = @id ? " id=\"#{@id}\"" : ""
      caption_html = @caption ? "<strong>Table #{number}:</strong> #{@caption}" : "<strong>Table #{number}</strong>"
      
      <<~HTML
        <div class="academic-table"#{id_attr} data-table-number="#{number}">
          <div class="table-caption">#{caption_html}</div>
          <div class="table-content">
            #{content}
          </div>
        </div>
      HTML
    end
  end

  # Equation block tag: {% equation id:eq-id %}...{% endequation %}
  class EquationBlock < Liquid::Block
    def initialize(tag_name, markup, tokens)
      super
      @id = nil
      
      if markup =~ /id:\s*([^\s]+)/
        @id = $1.strip
      end
    end

    def render(context)
      content = super.to_s.strip
      number = AcademicNumbering.increment('equation')
      
      if @id
        AcademicNumbering.register_label(@id, 'equation', number)
      end
      
      id_attr = @id ? " id=\"#{@id}\"" : ""
      
      # Wrap in equation environment with number
      <<~HTML
        <div class="academic-equation"#{id_attr} data-equation-number="#{number}">
          <div class="equation-content">
            $$#{content}$$
          </div>
          <span class="equation-number">(#{number})</span>
        </div>
      HTML
    end
  end

  # Algorithm block tag: {% algorithm id:alg-id caption:"Caption" %}...{% endalgorithm %}
  class AlgorithmBlock < Liquid::Block
    def initialize(tag_name, markup, tokens)
      super
      @id = nil
      @caption = nil
      
      if markup =~ /id:\s*([^\s]+)/
        @id = $1.strip
      end
      if markup =~ /caption:\s*"([^"]+)"/
        @caption = $1.strip
      elsif markup =~ /caption:\s*'([^']+)'/
        @caption = $1.strip
      end
    end

    def render(context)
      content = super.to_s.strip
      number = AcademicNumbering.increment('algorithm')
      
      if @id
        AcademicNumbering.register_label(@id, 'algorithm', number)
      end
      
      id_attr = @id ? " id=\"#{@id}\"" : ""
      caption_html = @caption ? "Algorithm #{number}: #{@caption}" : "Algorithm #{number}"
      
      <<~HTML
        <div class="academic-algorithm"#{id_attr} data-algorithm-number="#{number}">
          <div class="algorithm-header">#{caption_html}</div>
          <div class="algorithm-content">
            #{content}
          </div>
        </div>
      HTML
    end
  end

  # Reference tag: {% ref fig-id %} -> "Figure 1"
  class RefTag < Liquid::Tag
    def initialize(tag_name, markup, tokens)
      super
      @id = markup.strip
    end

    def render(context)
      # Output a placeholder that JavaScript will resolve
      # This handles the case where references appear before definitions
      "<a href=\"##{@id}\" class=\"ref-placeholder\" data-ref-id=\"#{@id}\">?</a>"
    end
  end

  # Equation reference tag: {% eqref eq-id %} -> "(1)"
  class EqRefTag < Liquid::Tag
    def initialize(tag_name, markup, tokens)
      super
      @id = markup.strip
    end

    def render(context)
      "<a href=\"##{@id}\" class=\"eqref-placeholder\" data-ref-id=\"#{@id}\">(?)</a>"
    end
  end

  # Hook to reset counters for each page
  Jekyll::Hooks.register :pages, :pre_render do |page|
    AcademicNumbering.reset_counters
  end

  Jekyll::Hooks.register :documents, :pre_render do |doc|
    AcademicNumbering.reset_counters
  end
end

Liquid::Template.register_tag('figure', Jekyll::FigureBlock)
Liquid::Template.register_tag('table', Jekyll::TableBlock)
Liquid::Template.register_tag('equation', Jekyll::EquationBlock)
Liquid::Template.register_tag('algorithm', Jekyll::AlgorithmBlock)
Liquid::Template.register_tag('ref', Jekyll::RefTag)
Liquid::Template.register_tag('eqref', Jekyll::EqRefTag)

File: assets/js/academic-numbering.js
/**
 * Academic Numbering System for Jekyll Blog Posts (GitHub Pages Compatible)
 * 
 * This script provides LaTeX-like automatic numbering and cross-referencing
 * for figures, tables, equations, and algorithms using pure HTML + JavaScript.
 * NO Jekyll plugins required - works on GitHub Pages!
 * 
 * Usage in Markdown posts:
 * 
 * 1. Figures:
 *    <figure class="academic-figure" id="fig-example">
 *      <img src="..." alt="...">
 *      <figcaption data-caption="An example figure"></figcaption>
 *    </figure>
 *    
 *    Reference: <a href="#fig-example" class="ref"></a>  -> "Figure 1"
 * 
 * 2. Tables:
 *    <div class="academic-table" id="tab-example" data-caption="Table caption">
 *      | Header | Data |
 *      |--------|------|
 *      | A      | B    |
 *    </div>
 *    
 *    Reference: <a href="#tab-example" class="ref"></a>  -> "Table 1"
 * 
 * 3. Equations (labeled):
 *    <div class="academic-equation" id="eq-energy">
 *      $$E = mc^2$$
 *    </div>
 *    
 *    Reference: <a href="#eq-energy" class="ref"></a> -> "Eq. 1"
 *    Or: <a href="#eq-energy" class="eqref"></a> -> "(1)"
 * 
 * 4. Algorithms:
 *    <div class="academic-algorithm" id="alg-sort" data-caption="Sorting Algorithm">
 *      ...pseudocode...
 *    </div>
 *    
 *    Reference: <a href="#alg-sort" class="ref"></a>  -> "Algorithm 1"
 */

(function() {
  'use strict';

  // ============================================================
  // Label Registry and Counters
  // ============================================================
  
  const labelRegistry = {};
  const counters = {
    figure: 0,
    table: 0,
    equation: 0,
    algorithm: 0
  };
  
  const prefixes = {
    figure: 'Figure',
    table: 'Table',
    equation: 'Eq.',
    algorithm: 'Algorithm'
  };

  /**
   * Reset counters for fresh processing
   */
  function resetCounters() {
    counters.figure = 0;
    counters.table = 0;
    counters.equation = 0;
    counters.algorithm = 0;
    Object.keys(labelRegistry).forEach(key => delete labelRegistry[key]);
  }

  /**
   * Process figures - add numbering and format captions
   */
  function processFigures() {
    const figures = document.querySelectorAll('.academic-figure');
    figures.forEach(fig => {
      counters.figure++;
      const num = counters.figure;
      
      fig.setAttribute('data-figure-number', num);
      
      if (fig.id) {
        labelRegistry[fig.id] = { type: 'figure', number: num, prefix: prefixes.figure };
      }
      
      // Format caption
      const figcaption = fig.querySelector('figcaption');
      if (figcaption) {
        const captionText = figcaption.getAttribute('data-caption') || figcaption.textContent;
        if (captionText && !figcaption.classList.contains('processed')) {
          figcaption.innerHTML = '<strong>Figure ' + num + ':</strong> ' + captionText;
          figcaption.classList.add('figure-caption', 'processed');
        }
      }
    });
  }

  /**
   * Process tables - add numbering and format captions
   */
  function processTables() {
    const tables = document.querySelectorAll('.academic-table');
    tables.forEach(tbl => {
      counters.table++;
      const num = counters.table;
      
      tbl.setAttribute('data-table-number', num);
      
      if (tbl.id) {
        labelRegistry[tbl.id] = { type: 'table', number: num, prefix: prefixes.table };
      }
      
      // Add caption if data-caption is present and not already processed
      const caption = tbl.getAttribute('data-caption');
      if (caption && !tbl.querySelector('.table-caption')) {
        const captionDiv = document.createElement('div');
        captionDiv.className = 'table-caption';
        captionDiv.innerHTML = '<strong>Table ' + num + ':</strong> ' + caption;
        tbl.insertBefore(captionDiv, tbl.firstChild);
      }
    });
  }

  /**
   * Process equations - add numbering
   */
  function processEquations() {
    const equations = document.querySelectorAll('.academic-equation');
    equations.forEach(eq => {
      counters.equation++;
      const num = counters.equation;
      
      eq.setAttribute('data-equation-number', num);
      
      if (eq.id) {
        labelRegistry[eq.id] = { type: 'equation', number: num, prefix: prefixes.equation };
      }
      
      // Add equation number if not already present
      if (!eq.querySelector('.equation-number')) {
        const numSpan = document.createElement('span');
        numSpan.className = 'equation-number';
        numSpan.textContent = '(' + num + ')';
        eq.appendChild(numSpan);
      }
    });
  }

  /**
   * Process algorithms - add numbering and format captions
   */
  function processAlgorithms() {
    const algorithms = document.querySelectorAll('.academic-algorithm');
    algorithms.forEach(alg => {
      counters.algorithm++;
      const num = counters.algorithm;
      
      alg.setAttribute('data-algorithm-number', num);
      
      if (alg.id) {
        labelRegistry[alg.id] = { type: 'algorithm', number: num, prefix: prefixes.algorithm };
      }
      
      // Add caption if data-caption is present and not already processed
      const caption = alg.getAttribute('data-caption');
      if (caption && !alg.querySelector('.algorithm-caption')) {
        const captionDiv = document.createElement('div');
        captionDiv.className = 'algorithm-caption';
        captionDiv.innerHTML = '<strong>Algorithm ' + num + ':</strong> ' + caption;
        alg.insertBefore(captionDiv, alg.firstChild);
      }
    });
  }

  /**
   * Resolve all cross-references
   */
  function resolveReferences() {
    // Process standard references: <a href="#id" class="ref"></a>
    document.querySelectorAll('a.ref[href^="#"]').forEach(el => {
      const refId = el.getAttribute('href').substring(1); // Remove #
      const label = labelRegistry[refId];
      
      if (label) {
        el.textContent = label.prefix + ' ' + label.number;
        el.classList.add('ref-link', 'ref-' + label.type);
      } else {
        el.textContent = '??';
        el.classList.add('ref-error');
        console.warn('Academic numbering: Unknown reference "' + refId + '"');
      }
    });
    
    // Process equation references: <a href="#id" class="eqref"></a> -> "(1)"
    document.querySelectorAll('a.eqref[href^="#"]').forEach(el => {
      const refId = el.getAttribute('href').substring(1);
      const label = labelRegistry[refId];
      
      if (label && label.type === 'equation') {
        el.textContent = '(' + label.number + ')';
        el.classList.add('eqref-link');
      } else {
        el.textContent = '(??)';
        el.classList.add('ref-error');
        console.warn('Academic numbering: Unknown equation reference "' + refId + '"');
      }
    });
  }

  // ============================================================
  // Initialization
  // ============================================================
  
  function initAcademicNumbering() {
    resetCounters();
    processFigures();
    processTables();
    processEquations();
    processAlgorithms();
    resolveReferences();
  }

  // Run after DOM is ready
  function onReady(fn) {
    if (document.readyState === 'loading') {
      document.addEventListener('DOMContentLoaded', fn);
    } else {
      fn();
    }
  }

  onReady(function() {
    // Wait for MathJax to finish if present
    if (window.MathJax && window.MathJax.startup && window.MathJax.startup.promise) {
      window.MathJax.startup.promise.then(function() {
        setTimeout(initAcademicNumbering, 50);
      });
    } else if (window.MathJax && window.MathJax.Hub) {
      // MathJax 2.x
      window.MathJax.Hub.Queue(function() {
        setTimeout(initAcademicNumbering, 50);
      });
    } else {
      // No MathJax or MathJax not loaded yet
      setTimeout(initAcademicNumbering, 100);
    }
  });

  // Expose for manual re-initialization
  window.academicNumbering = {
    init: initAcademicNumbering,
    registry: labelRegistry,
    counters: counters
  };

})();
